### PR TITLE
nixos/acme: Increase number of retries in testing

### PR DESCRIPTION
**Pull request by m1cr0man**
Helps to avoid failures in Hydra when the host server starts the web server too slowly.

###### Description of changes

These tests keep failing on Hydra, and without logs for
failed tests it has been hard to figure out what is going
wrong.

I ran 103 loops of the test suite on my own server and none
of them failed. This gave me some confidence it wasn't a simple
configuration error (like an incorrectly generated cert).

Next I reviewed these 103 loops and other successful build logs
on Hydra. I found that in cases where I was using openssl s_client
to validate the server certificates, the client would frequently have
to retry up to 3 times before a connection was established and the
test could pass. The retry limit is 3 so this has been quite close to failing
frequently. [Example failed test](https://hydra.nixos.org/build/201652934/nixlog/1)

```
(finished: waiting for unit acme-finished-c.example.test.target, in 0.04 seconds)
client: must succeed: openssl s_client -brief -verify 2 -CAfile /tmp/ca.crt -servername b.example.test -connect b.example.test:443 < /dev/null 2>&1
webserver # [   54.127193] nginx[1833]: 2022/12/11 13:18:54 [info] 1833#1833: *11 client closed connection while waiting for request, client: 192.168.1.2, server: 0.0.0.0:443
(finished: must succeed: openssl s_client -brief -verify 2 -CAfile /tmp/ca.crt -servername b.example.test -connect b.example.test:443 < /dev/null 2>&1, in 0.03 seconds)
client: must succeed: openssl s_client -brief -verify 2 -CAfile /tmp/ca.crt -servername c.example.test -connect c.example.test:443 < /dev/null 2>&1
webserver # [   54.151594] nginx[1833]: 2022/12/11 13:18:54 [info] 1833#1833: *12 client closed connection while waiting for request, client: 192.168.1.2, server: 0.0.0.0:443
(finished: must succeed: openssl s_client -brief -verify 2 -CAfile /tmp/ca.crt -servername c.example.test -connect c.example.test:443 < /dev/null 2>&1, in 0.02 seconds)
(finished: subtest: Runs 1 cert for account creation before others, in 0.12 seconds)
```

I can't see any other reason from these logs (nor another nice way to detect) that Nginx or openssl s_client would fail to handle these requests appropriately, other than maybe Nginx had not finished fully starting up. For this reason, I consider it effective enough to increase the retry count and retry delay for all functions relying on openssl + server connectivity.

###### Things done

Increased retries for applicable functions from 3 to 10, and increased retry delay from 3 to 6 seconds.

- Built on platform(s)
  - [X] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [X] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by mweinelt** [comment]
I don't have an issue with the reformatting, but this is like 95% reformatting and 5% the change mentioned in the commit message. So it should probably have a dedicated commit.

**Comment by m1cr0man** [comment]
Yeah it really messes with the git blame too and it has historically been quite useful for the ACME module. I will fix it this evening

**Comment by m1cr0man** [comment]
Sorry, I forgot to do this until now.. busy few days.

**Comment by m1cr0man** [comment]
@ofborg test acme

**Comment by RaitoBezarius** [comment]
My 2 cents but I think we should use exponential backoff here, under heavy load, the new values might not be enough and sometimes Hydra may encounter unexpected performance issues.

I don't know if we have global timeouts in the tasks, but if we do so, we should defer to them and let the scheduler do its work.

If we do not, I would rather have the problem be expressed as: the test should pass under 30 minutes.

Finally, if this PR is urgent to fix a practical problem, I have nothing against being merged as-is, but I would love to have a tracking issue of this kind of matter ("voodoo magic attempts with timeouts related to flaky tests in Hydra")

**Comment by m1cr0man** [comment]
Exponential backoff is a good idea, I'll see about implementing that. I can add some logic too for guessing the required wait time by persisting the sleep time on each round.

Wrt a global timeout, one does exist and additionally [one can be defined in meta too](https://nixos.org/manual/nixpkgs/stable/#var-meta-timeout). Thanks to folks on #nixos-dev on Matrix for pointing that out.

I'll implement this tonight and see how it goes.

**Comment by m1cr0man** [comment]
Set the timeout to 30 minutes and created a helper class for managing the backoff, sleeping + eventual failure.

**Comment by RaitoBezarius** [comment]
> Set the timeout to 30 minutes and created a helper class for managing the backoff, sleeping + eventual failure.

Beautiful <3, thank you a lot!

**Comment by m1cr0man** [comment]
@ofborg test acme

**Comment by RaitoBezarius** [comment]
I think it's good to merge, @m1cr0man anything left to do?

**Comment by m1cr0man** [comment]
Yeah it's good to go. I'm cautious of the new retry logic as miraculously it wasn't even required for that invocation I tried earlier today. I'd like to see it do at least one retry not on my local machine. I'll keep an eye on a few nixpkgs builds after it's merged anyway.

**Review by RaitoBezarius — APPROVED** [review]

**Pull request merged**
